### PR TITLE
[FLINK-21939][python] Support batch mode in Python DataStream API for process operation and reduce operation

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request support batch mode in Python DataStream API for process operation and reduce operation*

## Verifying this change

This change can be verified by tests test_reduce_function_without_data_types and test_reduce_with_state in BatchModeDataStreamTests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/datastream/tests/test_data_stream.py [hunk 1]
```diff
@@ -1068,6 +1013,35 @@ def on_timer(self, timestamp, ctx):
         expected_result.sort()
         self.assertEqual(expected_result, result)
 
+    def test_reduce_function_without_data_types(self):
```
Comment (SteNicholas) [inline flink-python/pyflink/datastream/tests/test_data_stream.py:1016]: Could this add `test_reduce_function_with_data_types` test case?

File: flink-python/pyflink/datastream/tests/test_data_stream.py [hunk 2]
```diff
@@ -1118,6 +1092,35 @@ def on_timer(self, timestamp, ctx):
...
```
Comment (SteNicholas) [inline flink-python/pyflink/datastream/tests/test_data_stream.py:1095]: Could this add `test_reduce_function_with_data_types` test case?

File: flink-python/pyflink/datastream/tests/test_data_stream.py (hunk 1 above)
Comment (dianfu) [reply flink-python/pyflink/datastream/tests/test_data_stream.py:1016]: I'll rename it to test_reduce_function to avoid confusion.

**Review by HuangXingBo — COMMENTED** [review]
This commit forgot to change the param name of `key_type_info` in `data_stream_job.py`.

File: docs/content/docs/dev/datastream/operators/overview.md [hunk 3]
```diff
@@ -756,4 +756,4 @@ someStream.filter(...).slotSharingGroup("name")
 some_stream.filter(...).slot_sharing_group("name")
 ```
 {{< /tab >}}
-{{< /tabs>}}
\ No newline at end of file
+{{< /tabs>}}
```
Comment (HuangXingBo) [inline docs/content/docs/dev/datastream/operators/overview.md:759]: unnecessary change

File: flink-python/pyflink/datastream/data_stream.py [hunk 4]
```diff
@@ -851,28 +851,43 @@ def reduce(self, func: Union[Callable, ReduceFunction]) -> 'DataStream':
                 raise TypeError("The input func must be a ReduceFunction or a callable function.")
         output_type = _from_java_type(self._original_data_type_info.get_java_type_info())
 
-        class KeyedReduceFunctionWrapper(KeyedProcessFunction):
+        class KeyedReduceProcessFunction(KeyedProcessFunction):
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/data_stream.py:854]: Maybe we can rename other wrapper functions together

File: flink-python/pyflink/datastream/functions.py [hunk 5]
```diff
@@ -758,9 +760,7 @@ class KeyedProcessFunction(Function, ABC):
 
     Note that access to keyed state and timers (which are also scoped to a key) is only available if
     the KeyedProcessFunction is applied on a KeyedStream.
-    """
 
-    """
     A function that process elements of a stream.
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/functions.py:764]: I think this useless comment also needs to be deleted.

**Review by HuangXingBo — COMMENTED** [review]
@dianfu Thanks a lot for the PR. Looks good overall. I only left a minor comments.

**Comment by dianfu** [comment]
@SteNicholas @HuangXingBo Thanks for your review. Have addressed your comments in the latest PR.

**Comment by dianfu** [comment]
@flinkbot run azure

**Review by HuangXingBo — APPROVED** [review]
@dianfu Thanks a lot for the update. LGTM.

**Review by SteNicholas — APPROVED** [review]
